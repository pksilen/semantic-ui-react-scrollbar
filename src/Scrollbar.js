// @flow
/* eslint-disable jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions */

import * as React from 'react'; // NOSONAR
import type { Element } from 'react';
import PropTypes from 'prop-types';
import { Icon } from 'semantic-ui-react';
import styles from './styles';

type Props = {
  changeScrollPosition: (number) => void,
  className: ?string,
  maxScrollPosition: number,
  orientation: 'horizontal' | 'vertical'
};

type State = {
  currentScrollPosition: number
};

export default class Scrollbar extends React.Component<Props, State> {
  static propTypes: { [key: $Keys<Props>]: any } = {
    changeScrollPosition: PropTypes.func.isRequired,
    className: PropTypes.string,
    maxScrollPosition: PropTypes.number.isRequired,
    orientation: PropTypes.oneOf(['horizontal', 'vertical'])
  };

  static defaultProps = {
    className: undefined,
    orientation: 'horizontal'
  };

  state = {
    currentScrollPosition: 0
  };

  scrollbarRef: { current: any };

  trackRef: { current: any };

  thumbRef: { current: any };

  trackWidth: number;

  trackLeft: number;

  thumbWidth: number;

  trackHeight: number;

  trackTop: number;

  thumbHeight: number;

  constructor(props: Props) {
    super(props);

    this.scrollbarRef = React.createRef();
    this.trackRef = React.createRef();
    this.thumbRef = React.createRef();

    this.trackWidth = 0;
    this.trackLeft = 0;
    this.thumbWidth = 0;

    this.trackHeight = 0;
    this.trackTop = 0;
    this.thumbHeight = 0;
  }

  componentDidMount() {
    this.trackWidth = this.trackRef.current.offsetWidth;
    this.trackLeft = this.trackRef.current.getBoundingClientRect().left;
    this.thumbWidth = this.thumbRef.current.offsetWidth;

    this.trackHeight = this.trackRef.current.offsetHeight;
    this.trackTop = this.trackRef.current.getBoundingClientRect().top;
    this.thumbHeight = this.thumbRef.current.offsetHeight;

    document.addEventListener('keydown', this.onKeyDown);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.onKeyDown);
  }

  onKeyDown = (keyboardEvent: KeyboardEvent) => {
    const { orientation } = this.props;

    if (window.getComputedStyle(this.scrollbarRef.current).getPropertyValue('visibility') === 'visible') {
      if (
        (orientation === 'horizontal' && keyboardEvent.code === 'ArrowLeft') ||
        (orientation === 'vertical' && keyboardEvent.code === 'ArrowUp')
      ) {
        this.scrollLeftOrUp();
      } else if (
        (orientation === 'horizontal' && keyboardEvent.code === 'ArrowRight') ||
        (orientation === 'vertical' && keyboardEvent.code === 'ArrowDown')
      ) {
        this.scrollRightOrDown();
      }
    }
  };

  updateThumbPosition = (event: SyntheticMouseEvent<HTMLElement>) => {
    event.stopPropagation();

    const { changeScrollPosition, maxScrollPosition, orientation } = this.props;
    let currentScrollPosition;

    if (orientation === 'horizontal') {
      currentScrollPosition = Math.round(
        ((event.pageX - this.trackLeft - this.thumbWidth / 2) / (this.trackWidth - this.thumbWidth)) *
          maxScrollPosition
      );
    } else {
      currentScrollPosition = Math.round(
        ((event.pageY - this.trackTop - this.thumbHeight / 2) / (this.trackHeight - this.thumbHeight)) *
          maxScrollPosition
      );
    }

    currentScrollPosition = Math.min(maxScrollPosition, Math.max(0, currentScrollPosition));

    this.setState({
      currentScrollPosition
    });

    changeScrollPosition(currentScrollPosition);
  };

  endThumbDrag = (event: SyntheticMouseEvent<HTMLElement>) => {
    event.stopPropagation();
    window.removeEventListener('mousemove', this.updateThumbPosition);
    window.removeEventListener('mouseup', this.endThumbDrag);
  };

  startThumbDrag = (event: SyntheticMouseEvent<HTMLElement>) => {
    event.stopPropagation();
    window.addEventListener('mousemove', this.updateThumbPosition);
    window.addEventListener('mouseup', this.endThumbDrag);
  };

  scrollLeftOrUp = () => {
    const { changeScrollPosition } = this.props;

    this.setState(({ currentScrollPosition }: State): State => {
      if (currentScrollPosition > 0) {
        changeScrollPosition(currentScrollPosition - 1);
        return {
          currentScrollPosition: currentScrollPosition - 1
        };
      }

      return {
        currentScrollPosition
      };
    });
  };

  scrollRightOrDown = () => {
    const { maxScrollPosition, changeScrollPosition } = this.props;

    this.setState(({ currentScrollPosition }: State): State => {
      if (currentScrollPosition < maxScrollPosition) {
        changeScrollPosition(currentScrollPosition + 1);
        return {
          currentScrollPosition: currentScrollPosition + 1
        };
      }

      return {
        currentScrollPosition
      };
    });
  };

  render(): Element<*> {
    const { className, maxScrollPosition, orientation } = this.props;
    const { currentScrollPosition } = this.state;

    let marginLeft = 0;
    let marginTop = 0;

    if (orientation === 'horizontal') {
      marginLeft = `${
        this.trackWidth
          ? (currentScrollPosition * (this.trackWidth - this.thumbWidth)) / maxScrollPosition
          : 0
      }px`;
    } else {
      marginTop = `${
        this.trackHeight
          ? (currentScrollPosition * (this.trackHeight - this.thumbHeight)) / maxScrollPosition
          : 0
      }px`;
    }

    const thumbStyle = {
      ...styles.thumb,
      marginLeft,
      marginTop
    };

    return (
      <div className={className} ref={this.scrollbarRef} style={styles.scrollbar[orientation]}>
        <div className="button" style={styles.button} onClick={this.scrollLeftOrUp}>
          <Icon name={orientation === 'horizontal' ? 'caret left' : 'caret up'} style={{ margin: 0 }} />
        </div>
        <div
          className="track"
          style={styles.track[orientation]}
          ref={this.trackRef}
          onClick={this.updateThumbPosition}
        >
          <div className="thumb" style={thumbStyle} ref={this.thumbRef} onMouseDown={this.startThumbDrag} />
        </div>
        <div className="button" style={styles.button} onClick={this.scrollRightOrDown}>
          <Icon name={orientation === 'horizontal' ? 'caret right' : 'caret down'} style={{ margin: 0 }} />
        </div>
      </div>
    );
  }
}
