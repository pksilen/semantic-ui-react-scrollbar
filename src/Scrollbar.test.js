import React from 'react';
import { mount } from 'enzyme';
import PropTypes from 'prop-types';
import Scrollbar from './Scrollbar';

let changeScrollPositionMock;

beforeEach(() => {
  changeScrollPositionMock = jest.fn();
});

describe('propTypes', () => {
  it('should define prop types', () => {
    // eslint-disable-next-line react/forbid-foreign-prop-types
    const { propTypes } = Scrollbar;

    expect(propTypes.changeScrollPosition).toBe(PropTypes.func.isRequired);
    expect(propTypes.className).toBe(PropTypes.string);
    expect(propTypes.maxScrollPosition).toBe(PropTypes.number.isRequired);
    expect(propTypes.orientation).toBeTruthy();
  });
});

describe('defaultProps', () => {
  it('should set default prop values', () => {
    const scrollbarWrapper = mount(
      <Scrollbar maxScrollPosition={10} changeScrollPosition={changeScrollPositionMock} />
    );

    expect(scrollbarWrapper.props().className).toBe(undefined);
    expect(scrollbarWrapper.props().orientation).toBe('horizontal');
  });
});

describe('onKeyDown', () => {
  test.each([
    ['horizontal', 'ArrowLeft', 'scrollLeftOrUp'],
    ['horizontal', 'ArrowRight', 'scrollRightOrDown'],
    ['vertical', 'ArrowUp', 'scrollLeftOrUp'],
    ['vertical', 'ArrowDown', 'scrollRightOrDown']
  ])('it should scroll scroll bar when arrow key is pressed', (orientation, keyCode, functionName) => {
    const scrollbarWrapper = mount(
      <Scrollbar
        maxScrollPosition={10}
        changeScrollPosition={changeScrollPositionMock}
        orientation={orientation}
      />
    );
    const scrollbar = scrollbarWrapper.instance();
    scrollbar[functionName] = jest.fn();
    scrollbar.scrollbarRef.current.style.visibility = 'visible';

    const keyDownEvent = new KeyboardEvent('keydown', { code: keyCode });
    global.document.dispatchEvent(keyDownEvent);

    expect(scrollbar[functionName]).toHaveBeenCalledTimes(1);
  });
});

describe('updateThumbPosition', () => {
  it('should update current scroll position for horizontal scroll bar', () => {
    const scrollbarWrapper = mount(
      <Scrollbar maxScrollPosition={10} changeScrollPosition={changeScrollPositionMock} />
    );

    const scrollbar = scrollbarWrapper.instance();
    scrollbar.trackWidth = 100;
    scrollbar.thumbWidth = 5;
    scrollbar.trackLeft = 0;

    const track = scrollbarWrapper.find('.track');
    track.simulate('click', { pageX: 50 });

    expect(scrollbarWrapper.state('currentScrollPosition')).toBe(5);
  });

  it('should update current scroll position for vertical scroll bar', () => {
    const scrollbarWrapper = mount(
      <Scrollbar
        maxScrollPosition={10}
        changeScrollPosition={changeScrollPositionMock}
        orientation="vertical"
      />
    );
    const scrollbar = scrollbarWrapper.instance();
    scrollbar.trackHeight = 100;
    scrollbar.thumbWidth = 5;
    scrollbar.trackTop = 0;

    const track = scrollbarWrapper.find('.track');
    track.simulate('click', { pageY: 50 });

    expect(scrollbarWrapper.state('currentScrollPosition')).toBe(5);
  });
});

describe('startThumbDrag', () => {
  it('should update current scroll position', () => {
    const scrollbarWrapper = mount(
      <Scrollbar maxScrollPosition={10} changeScrollPosition={changeScrollPositionMock} />
    );
    const scrollbar = scrollbarWrapper.instance();
    scrollbar.trackWidth = 100;
    scrollbar.thumbWidth = 5;
    scrollbar.trackLeft = 0;
    const thumb = scrollbarWrapper.find('.thumb');

    thumb.simulate('mousedown');

    const mouseMoveEvent = new MouseEvent('mousemove');
    mouseMoveEvent.pageX = 50;
    global.window.dispatchEvent(mouseMoveEvent);

    const mouseUpEvent = new MouseEvent('mouseup');
    global.window.dispatchEvent(mouseUpEvent);

    expect(scrollbarWrapper.state('currentScrollPosition')).toBe(5);
  });
});

describe('scrollLeftOrUp', () => {
  it('should scroll scroll bar left or up', () => {
    const scrollbarWrapper = mount(
      <Scrollbar maxScrollPosition={10} changeScrollPosition={changeScrollPositionMock} />
    );

    const firstButton = scrollbarWrapper.find('.button').first();
    const lastButton = scrollbarWrapper.find('.button').last();
    lastButton.simulate('click');
    firstButton.simulate('click');

    expect(scrollbarWrapper.state('currentScrollPosition')).toBe(0);
  });

  it('should not scroll scroll bar left or up if scroll bar is at leftmost or topmost position', () => {
    const scrollbarWrapper = mount(
      <Scrollbar maxScrollPosition={10} changeScrollPosition={changeScrollPositionMock} />
    );

    const firstButton = scrollbarWrapper.find('.button').first();
    firstButton.simulate('click');

    expect(scrollbarWrapper.state('currentScrollPosition')).toBe(0);
  });
});

describe('scrollRightOrDown', () => {
  it('should scroll scroll bar left or up', () => {
    const scrollbarWrapper = mount(
      <Scrollbar maxScrollPosition={10} changeScrollPosition={changeScrollPositionMock} />
    );

    const lastButton = scrollbarWrapper.find('.button').last();
    lastButton.simulate('click');

    expect(scrollbarWrapper.state('currentScrollPosition')).toBe(1);
  });

  it('should not scroll scroll bar right or down if scroll bar is at rightmost or bottommost position', () => {
    const scrollbarWrapper = mount(
      <Scrollbar maxScrollPosition={1} changeScrollPosition={changeScrollPositionMock} />
    );

    const lastButton = scrollbarWrapper.find('.button').last();
    lastButton.simulate('click');
    lastButton.simulate('click');

    expect(scrollbarWrapper.state('currentScrollPosition')).toBe(1);
  });
});

describe('render', () => {
  it('should render horizontal scroll bar correctly', () => {
    const scrollbarWrapper = mount(
      <Scrollbar maxScrollPosition={10} changeScrollPosition={changeScrollPositionMock} />
    );

    expect(scrollbarWrapper).toMatchSnapshot();

    scrollbarWrapper.unmount();
  });

  it('should render vertical scroll bar correctly', () => {
    const scrollbarWrapper = mount(
      <Scrollbar
        maxScrollPosition={10}
        changeScrollPosition={changeScrollPositionMock}
        orientation="vertical"
      />
    );

    expect(scrollbarWrapper).toMatchSnapshot();
  });
});
