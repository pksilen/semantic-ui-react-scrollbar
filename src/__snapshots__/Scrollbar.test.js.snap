// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render horizontal scroll bar correctly 1`] = `
<Scrollbar
  changeScrollPosition={[MockFunction]}
  maxScrollPosition={10}
  orientation="horizontal"
>
  <div
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "display": "flex",
        "left": 0,
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <div
      className="button"
      onClick={[Function]}
      style={
        Object {
          "flexShrink": 0,
          "fontSize": "1.2rem",
          "lineHeight": "1.2rem",
        }
      }
    >
      <Icon
        as="i"
        name="caret left"
        style={
          Object {
            "margin": 0,
          }
        }
      >
        <i
          aria-hidden="true"
          className="caret left icon"
          onClick={[Function]}
          style={
            Object {
              "margin": 0,
            }
          }
        />
      </Icon>
    </div>
    <div
      className="track"
      onClick={[Function]}
      style={
        Object {
          "background": "rgba(0, 0, 0, 0.1)",
          "flexGrow": 1,
          "height": "0.8rem",
        }
      }
    >
      <div
        className="thumb"
        onMouseDown={[Function]}
        style={
          Object {
            "background": "rgb(0, 0, 0)",
            "height": "0.8rem",
            "marginLeft": "0px",
            "marginTop": 0,
            "width": "0.8rem",
          }
        }
      />
    </div>
    <div
      className="button"
      onClick={[Function]}
      style={
        Object {
          "flexShrink": 0,
          "fontSize": "1.2rem",
          "lineHeight": "1.2rem",
        }
      }
    >
      <Icon
        as="i"
        name="caret right"
        style={
          Object {
            "margin": 0,
          }
        }
      >
        <i
          aria-hidden="true"
          className="caret right icon"
          onClick={[Function]}
          style={
            Object {
              "margin": 0,
            }
          }
        />
      </Icon>
    </div>
  </div>
</Scrollbar>
`;

exports[`render should render vertical scroll bar correctly 1`] = `
<Scrollbar
  changeScrollPosition={[MockFunction]}
  maxScrollPosition={10}
  orientation="vertical"
>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "height": "100%",
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <div
      className="button"
      onClick={[Function]}
      style={
        Object {
          "flexShrink": 0,
          "fontSize": "1.2rem",
          "lineHeight": "1.2rem",
        }
      }
    >
      <Icon
        as="i"
        name="caret up"
        style={
          Object {
            "margin": 0,
          }
        }
      >
        <i
          aria-hidden="true"
          className="caret up icon"
          onClick={[Function]}
          style={
            Object {
              "margin": 0,
            }
          }
        />
      </Icon>
    </div>
    <div
      className="track"
      onClick={[Function]}
      style={
        Object {
          "background": "rgba(0, 0, 0, 0.1)",
          "flexGrow": 1,
          "width": "0.8rem",
        }
      }
    >
      <div
        className="thumb"
        onMouseDown={[Function]}
        style={
          Object {
            "background": "rgb(0, 0, 0)",
            "height": "0.8rem",
            "marginLeft": 0,
            "marginTop": "0px",
            "width": "0.8rem",
          }
        }
      />
    </div>
    <div
      className="button"
      onClick={[Function]}
      style={
        Object {
          "flexShrink": 0,
          "fontSize": "1.2rem",
          "lineHeight": "1.2rem",
        }
      }
    >
      <Icon
        as="i"
        name="caret down"
        style={
          Object {
            "margin": 0,
          }
        }
      >
        <i
          aria-hidden="true"
          className="caret down icon"
          onClick={[Function]}
          style={
            Object {
              "margin": 0,
            }
          }
        />
      </Icon>
    </div>
  </div>
</Scrollbar>
`;
